### ISSUE TITLE
[Ruby] Typed metavariable handling new, literal arrays

### ISSUE THREAD

**Issue by dlukeomalley** [semgrep/semgrep#1301]
https://semgrep.dev/s/ExwN?version=develop

 Doesn’t catch the Array.new from either casting of syntax

**Comment by aryx** [comment]
Not sure what you want, but your semgrep.live link above has the wrong pattern.
You need $ARR.instance_eval(...) as in https://semgrep.dev/s/K39j/?version=develop

**Comment by aryx** [comment]
if you're talking about the alias of 'foo' created when called Array.new(), which might be an alias for 'Array' (not super familiar with Ruby), but yes it's not handled by semgrep. You need $ARR.instance_eval or we need some kind of typed metavariables for Ruby to handle this case.

**Comment by aryx** [comment]
But the title of this issue is confusing. Maybe "support typed metavariable for Ruby" or "handle type aliases".

**Comment by dlukeomalley** [comment]
> Not sure what you want, but your semgrep.live link above has the wrong pattern.

It's the right pattern for the behavior I _want_ Semgrep to have, but is not right in the context of how Semgrep actually works. 

> we need some kind of typed metavariables

Thanks for the clarification. It makes sense to me now that what I want to query for is "instance_eval calls on objects of type Array", which isn't supported.

@nobletrout - Does it suffice for you to use `$ARR.instance_eval` for now?

**Comment by aryx** [comment]
A nice example of "complex symbolic expression" constant propagation (or typed metavariable in Ruby).

**Comment by aryx** [comment]
keep it

**Comment by IagoAbal** [comment]
I think this is more about typing. To make those two examples work maybe it's not too difficult, but in general we would need a type inferrer for the generic AST. I guess we focus on those two simple cases for now?

**Comment by aryx** [comment]
yep, just make the basic examples work should be enough for now. I hope the work on symbolic expression propagation
can get those for free.

**Comment by aryx** [comment]
:wave:

**Comment by IagoAbal** [comment]
Another thing that would be great to get done in Q4.

**Comment by emjin** [comment]
+1

**Comment by Sjord** [comment]
Links were broken, here are working links:

* https://semgrep.dev/s/ExwN?version=develop
* https://semgrep.dev/s/K39j/?version=develop

The title indicates Ruby, but this would be useful in any object-oriented language. For example, in PHP I want to match calls to `DOMDocument::load(...)`, which would look something like this:

```
$doc = new DOMDocument();
$doc->load('book.xml');
```

So I want semgrep to remember the type of `$doc`, so that it later knows that the call to `load()` is actually to `DOMDocument::load`.

**Comment by IagoAbal** [comment]
We have implemented "symbolic propagation" now (#4388), but I think all the examples above are better solved with typed metavariables.

**Comment by IagoAbal** [comment]
@Sjord's example should work but I think it doesn't mainly because `new DOMDocument()` is not being correctly translated into Generic (`DOMDocument` should be an `ArgType`).

**Comment by ievans** [comment]
@dlukeomalley this works now! See https://semgrep.dev/s/q9Bx

**Comment by aryx** [comment]
I don't think you posted the right link isaac. The thing you wrote has been supported for a very long time.\

**Comment by IagoAbal** [comment]
Yeah the problem is still not solved. And I think it just requires better type inference/propagation in Generic or for Ruby/PHP in particular. Symbolic propagation of `Array.new()` can handle one of the cases, but not the rest.

**Comment by aryx** [comment]
it will takes us 2 years to fulfill the vision of luke :)

**Comment by nobletrout** [comment]
I have been following this issue since its inception. It's become my favorite issue to watch open/close repeatedly. :)

**Comment by aryx** [comment]
lol

**Comment by IagoAbal** [comment]
Maybe something we could do for the next hackathon? cc @mmcqd
